homeassistant/components/roborock/roborock_storage.py

```diff
@@ -0,0 +1,152 @@
+            self._data[map_name] = RoborockMapEntry(
+                map_name,
+                dt_util.utcnow().timestamp(),
+            )
+            results.append(map_data)
+        return results
+
+    def _save_map(self, filename: str, content: bytes) -> None:
+        """Help other functions save the map. Should not be called separately."""
+        os.makedirs(os.path.dirname(filename), exist_ok=True)
+        _LOGGER.debug("Saving event map to disk store: %s", filename)
+        with open(filename, "wb") as stored_map:
+            stored_map.write(content)
+
+    async def async_save_map(self, map_name: str, content: bytes) -> None:
+        """Write map if it should be updated."""
+        await self.async_save_maps([(map_name, content)])
+
+    async def async_save_maps(self, maps: list[tuple[str, bytes]]) -> None:
+        """Write maps - update regardless. Should be called as background task."""
+
+        def save_maps(maps: list[tuple[str, bytes]]) -> None:
+            for map_name, content in maps:
+                map_entry = self._data.get(map_name)
+                filename = self._get_map_filename(map_name)
+                self._data[map_name] = RoborockMapEntry(
+                    map_name, dt_util.utcnow().timestamp()
+                )
+                try:
+                    self._save_map(filename, content)
+                except OSError as err:
+                    _LOGGER.error("Unable to write map file: %s %s", filename, err)
+                    # We don't want the _data dict to be updated with incorrect information.
+                    if map_entry is not None:
+                        self._data[map_name] = map_entry
+
+        await self._hass.async_add_executor_job(save_maps, maps)
+
+    async def async_remove_maps(self) -> None:
+        """Remove all maps associated with a config entry."""
+
+        def remove_maps() -> None:
+            directory = self._hass.config.path(f"{MAP_PATH}/{self._entry_id}")
+            try:
+                for filename in os.listdir(directory):
+                    file_path = os.path.join(directory, filename)
+                    _LOGGER.debug("Removing map from disk store: %s", file_path)
+                    os.remove(file_path)
+            except OSError as err:
+                _LOGGER.error(
+                    "Unable to remove map files for: %s %s", self._entry_id, err
+                )
+
+        await self._hass.async_add_executor_job(remove_maps)
+
+    async def async_save_map_info(
+        self, coord_duid: str, map_info: dict[int, RoborockMapInfo]
+    ) -> None:
+        """Save the coordinator map info."""
+        self._map_info[coord_duid] = map_info
+        await self._store.async_save(self._map_info)
+
+    async def async_load_map_info(self, coord_duid: str) -> dict[int, RoborockMapInfo]:
+        """Load a coordinator's map info."""
+        map_info: dict[str, dict[str, dict]] | None = await self._store.async_load()
```

Comment:
This reads from disk every time load is called so the use of `self._map_info` does not seem to be adding value. Is the idea to treat `_map_info` as a cache or should it be dropped entirely? It seems like the intent was to load once at startup, then always read from it and write through directly to disk every time.

Reply:
This is only called once. _map_info is a cache so that we can update the coord that changed and then rewrite it.

The coord that calls the save does not have access to all of the data, so roborock storage holds it instead

Reply:
That would mean I'm doubling the IO everytime I am writing to the disk - I can do that but it seems a odd. Happy to do it if you think that is best

Reply:
I assumed updates were rare, so i was trying to see about removing complexity. If this is really a performance sensitive flow then it could also be improved by not storing all the maps in a single file since updating maps for one `coord_duid` requires updating all of them.

If we want to make this work, then there are some either potential bugs here or semantics that need to be enforced in how this is used:
- If you call save before calling load then the existing map data gets overwritten
- Calling remove map info doesn't clear the map info so then calling save will write back old map data

These work today because of assumptions in exactly how these are used by the callers today, which i think make this a little brittle, or spread logic across two files.

Removing `_map_info` sidesteps needing to address these bugs / or add documentation explaining assumptions / or add invariant checking to enforce a specific call pattern.

Reply:
I think what makes this fancy is that its trying to do two things (1) provide a cache of the entire disk and (2) provide partial update semantics with read/write of a subset of the entire cached data. 

Separating the responsibilities and having the cache at a lower level, reading /writing the entire dataset would make the semantics simpler and easier to reason about. (e.g. cache the entire data on load, overwrite the entire data on save.

But more concretely i think these fixes can make what is here work:
- only ever read from disk once, always prefer to read from the existing data if it was already read
- invalidate map info when removing it from disk
- consider using `async_delay_save` on write to batch writes together from multiple devices